// @react/react-spectrum/Grid/js/GridRow.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classnames = _interopRequireDefault(require("classnames"));

var _filterDOMProps = _interopRequireDefault(require("../../utils/filterDOMProps"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _responsive = _interopRequireDefault(require("./responsive"));

/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2019 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.
**************************************************************************/

/**
 * A row in a grid
 */
var GridRow =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(GridRow, _Component);

  function GridRow() {
    (0, _classCallCheck2["default"])(this, GridRow);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(GridRow).apply(this, arguments));
  }

  (0, _createClass2["default"])(GridRow, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          align = _this$props.align,
          children = _this$props.children,
          valign = _this$props.valign,
          distribution = _this$props.distribution,
          reverse = _this$props.reverse,
          className = _this$props.className,
          otherProps = (0, _objectWithoutProperties2["default"])(_this$props, ["align", "children", "valign", "distribution", "reverse", "className"]);
      var classes = (0, _classnames["default"])('spectrum-grid-row', (0, _responsive["default"])('spectrum-grid-#value-#size', align), (0, _responsive["default"])('spectrum-grid-#value-#size', valign), (0, _responsive["default"])('spectrum-grid-#value-#size', distribution), {
        'spectrum-grid-reverse': reverse
      }, className);
      return _react["default"].createElement("div", (0, _extends2["default"])({
        className: classes
      }, (0, _filterDOMProps["default"])(otherProps)), children);
    }
  }]);
  return GridRow;
}(_react.Component);

exports["default"] = GridRow;
(0, _defineProperty2["default"])(GridRow, "propTypes", {
  /**
   * Fill me in and fix my prop type
   */
  align: _propTypes["default"].any,

  /**
   * Fill me in and fix my prop type
   */
  valign: _propTypes["default"].any,

  /**
   * Fill me in and fix my prop type
   */
  distribution: _propTypes["default"].any,

  /**
   * Fill me in and fix my prop type
   */
  reverse: _propTypes["default"].any,

  /**
   * Fill me in and fix my prop type
   */
  className: _propTypes["default"].string
});
(0, _defineProperty2["default"])(GridRow, "defaultProps", {
  reverse: false
});
