// @react/react-spectrum/ColumnView/js/Item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _Checkbox = _interopRequireDefault(require("../../Checkbox"));

var _ChevronRightMedium = _interopRequireDefault(require("../../Icon/core/ChevronRightMedium"));

var _classnames = _interopRequireDefault(require("classnames"));

var _convertUnsafeMethod = _interopRequireDefault(require("../../utils/convertUnsafeMethod"));

var _filterDOMProps = _interopRequireDefault(require("../../utils/filterDOMProps"));

var _focusRing = _interopRequireDefault(require("../../utils/focusRing"));

var _react = _interopRequireDefault(require("react"));

var _class;

var Item =
/*
 * A wrapper for an Item within an ItemColumn that will manage the Item's state
 */
(0, _convertUnsafeMethod["default"])(_class = (0, _focusRing["default"])(_class =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(Item, _React$Component);

  function Item() {
    (0, _classCallCheck2["default"])(this, Item);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Item).apply(this, arguments));
  }

  (0, _createClass2["default"])(Item, [{
    key: "focus",
    value: function focus() {
      if (this.itemRef) {
        this.itemRef.focus();
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props = this.props,
          item = _this$props.item,
          renderItem = _this$props.renderItem,
          highlighted = _this$props.selected,
          allowsSelection = _this$props.allowsSelection,
          allowsBranchSelection = _this$props.allowsBranchSelection,
          isSelected = _this$props.isSelected,
          focused = _this$props.focused,
          column = _this$props.column,
          _this$props$level = _this$props.level,
          level = _this$props$level === void 0 ? 0 : _this$props$level,
          detailNode = _this$props.detailNode,
          collectionView = _this$props.collectionView,
          props = (0, _objectWithoutProperties2["default"])(_this$props, ["item", "renderItem", "selected", "allowsSelection", "allowsBranchSelection", "isSelected", "focused", "column", "level", "detailNode", "collectionView"]);
      var className = (0, _classnames["default"])('spectrum-AssetList-item', {
        'is-selectable': allowsBranchSelection,
        'is-branch': item.hasChildren,
        'is-navigated': highlighted && !isSelected,
        'is-selected': isSelected
      });
      var id = item.getItemId();
      var labelId = "".concat(id, "-label");
      var columnFocused = column && column.props.focused;
      var tabIndex = columnFocused && focused ? 0 : -1;

      if (columnFocused && !focused && collectionView && !collectionView.focusedIndexPath) {
        tabIndex = 0;
      }

      var setSize = item.parent.children.getSectionLength(0);
      var isExpanded = item.hasChildren ? highlighted || false : null;
      var ownedColumnId = isExpanded || detailNode === item ? item.getColumnId() : null;
      var ariaDescribedby = detailNode === item ? ownedColumnId : null;
      return _react["default"].createElement("div", (0, _extends2["default"])({
        ref: function ref(i) {
          return _this.itemRef = i;
        },
        id: id,
        tabIndex: tabIndex,
        role: "treeitem",
        "aria-selected": allowsSelection && (allowsBranchSelection || !item.hasChildren) ? isSelected || false : null,
        "aria-level": level + 1,
        "aria-posinset": item.index + 1,
        "aria-setsize": setSize,
        "aria-expanded": isExpanded,
        "aria-labelledby": labelId,
        "aria-describedby": ariaDescribedby,
        "aria-owns": ownedColumnId,
        className: className
      }, (0, _filterDOMProps["default"])(props)), allowsSelection && allowsBranchSelection && this.renderCheckbox(labelId), _react["default"].createElement("span", {
        role: "presentation",
        className: "spectrum-AssetList-itemLabel",
        id: labelId
      }, renderItem(item.item, item)), item.hasChildren && this.renderChevron(), allowsSelection && !allowsBranchSelection && !item.hasChildren && this.renderCheckbox());
    }
  }, {
    key: "renderChevron",
    value: function renderChevron() {
      return _react["default"].createElement(_ChevronRightMedium["default"], {
        className: "spectrum-AssetList-itemChildIndicator",
        size: "XS"
      });
    }
  }, {
    key: "renderCheckbox",
    value: function renderCheckbox(labelId) {
      var _this2 = this;

      var _this$props2 = this.props,
          isSelected = _this$props2.isSelected,
          onSelect = _this$props2.onSelect;
      return _react["default"].createElement(_Checkbox["default"], {
        "aria-hidden": "true",
        "aria-labelledby": labelId,
        tabIndex: -1,
        className: "spectrum-AssetList-itemSelector",
        onMouseDown: function onMouseDown(e) {
          return e.stopPropagation();
        },
        onFocus: function onFocus(e) {
          e.preventDefault();

          _this2.focus();
        },
        checked: isSelected,
        onChange: onSelect
      });
    }
  }]);
  return Item;
}(_react["default"].Component)) || _class) || _class;

exports["default"] = Item;
