// @react/react-spectrum/StepList/js/StepList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = StepList;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _TabListBase = _interopRequireDefault(require("../../TabList/js/TabListBase"));

/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2019 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.
**************************************************************************/
if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/steplist/index.css");

  require("../../spectrum-css/steplist/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/steplist/index-lg.css");
} else {
  require("../../spectrum-css/steplist/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/steplist/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/steplist/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/steplist/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/steplist/multiStops/darkest.css");
var SMALL = 'S';
var LARGE = 'L';
var ON = 'on';

function StepList(_ref) {
  var className = _ref.className,
      _ref$interaction = _ref.interaction,
      interaction = _ref$interaction === void 0 ? ON : _ref$interaction,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? LARGE : _ref$size,
      otherProps = (0, _objectWithoutProperties2["default"])(_ref, ["className", "interaction", "size"]);
  return _react["default"].createElement(_TabListBase["default"], (0, _extends2["default"])({}, otherProps, {
    className: (0, _classnames["default"])('spectrum-Steplist', {
      'spectrum-Steplist--interactive': interaction === ON,
      'spectrum-Steplist--small': size === SMALL
    }, className),
    disabled: interaction !== ON,
    "aria-multiselectable": false,
    childMappingFunction: function childMappingFunction() {
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return getChildProps.apply(void 0, [size].concat(args));
    }
  }));
}

function getChildProps(size, tabList, child, index) {
  return {
    complete: +tabList.state.selectedIndex > index,
    size: size
  };
}

StepList.displayName = 'StepList';
StepList.propTypes = {
  /** Class to add to the step list */
  className: _propTypes["default"].string,

  /** Index of the selected step (uncontrolled state) */
  defaultSelectedIndex: _propTypes["default"].number,

  /** Turn step list interaction (clickable to progress) on or off */
  interaction: _propTypes["default"].oneOf([ON, 'off']),

  /** Index of the selected step */
  selectedIndex: _propTypes["default"].number,

  /** Size of the steps and step list */
  size: _propTypes["default"].oneOf(['S', 'L'])
};
StepList.defaultProps = {
  interaction: ON,
  size: LARGE
};
