// @react/react-spectrum/StepList/js/Step.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Step;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classnames = _interopRequireDefault(require("classnames"));

var _filterDOMProps = _interopRequireDefault(require("../../utils/filterDOMProps"));

var _OverlayTrigger = _interopRequireDefault(require("../../OverlayTrigger"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Tooltip = _interopRequireDefault(require("../../Tooltip"));

/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2019 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.
**************************************************************************/
var SMALL = 'S';

function Step(_ref) {
  var children = _ref.children,
      className = _ref.className,
      complete = _ref.complete,
      selected = _ref.selected,
      size = _ref.size,
      otherProps = (0, _objectWithoutProperties2["default"])(_ref, ["children", "className", "complete", "selected", "size"]);
  return _react["default"].createElement("div", (0, _extends2["default"])({
    className: (0, _classnames["default"])(className, 'spectrum-Steplist-item', {
      'is-complete': complete
    }, {
      'is-selected': selected
    }),
    role: "tab",
    "aria-selected": selected
  }, (0, _filterDOMProps["default"])(otherProps)), _react["default"].createElement("div", {
    className: "spectrum-Steplist-label"
  }, size !== SMALL && children), _react["default"].createElement("span", {
    className: "spectrum-Steplist-markerContainer"
  }, size === SMALL ? _react["default"].createElement(_OverlayTrigger["default"], {
    placement: "top"
  }, _react["default"].createElement("span", {
    className: "spectrum-Steplist-marker"
  }), _react["default"].createElement(_Tooltip["default"], {
    open: true,
    className: "spectrum-Steplist-tooltip"
  }, children)) : _react["default"].createElement("span", {
    className: "spectrum-Steplist-marker"
  })), _react["default"].createElement("span", {
    className: "spectrum-Steplist-segment"
  }));
}

Step.displayName = 'Step';
Step.propTypes = {
  /** Class to add to the step */
  className: _propTypes["default"].string,

  /** Whether the current step is completed */
  complete: _propTypes["default"].bool,

  /** Whether the current step is selected */
  selected: _propTypes["default"].bool,

  /** Define the size of the step */
  size: _propTypes["default"].oneOf(['S', 'L'])
};
