// @react/react-spectrum/ComboBox/js/ComboBox.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault3(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault3(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault3(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault3(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault3(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault3(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault3(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault3(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault3(require("@babel/runtime/helpers/defineProperty"));

var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));

var _autobindDecorator = _interopRequireDefault3(require("autobind-decorator"));

var _Autocomplete = _interopRequireDefault3(require("../../Autocomplete"));

var _Button = _interopRequireDefault3(require("../../Button"));

var _events = require("../../utils/events");

var _ChevronDownMedium = _interopRequireDefault3(require("../../Icon/core/ChevronDownMedium"));

var _classnames = _interopRequireDefault3(require("classnames"));

var _convertUnsafeMethod = _interopRequireDefault3(require("../../utils/convertUnsafeMethod"));

var _createId = _interopRequireDefault3(require("../../utils/createId"));

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault3(require("prop-types"));

var _react = _interopRequireDefault3(require("react"));

var _Textfield = _interopRequireDefault3(require("../../Textfield"));

var _class, _class2, _temp;

var intlMessages = {
  "cs-CZ": (0, _interopRequireDefault2["default"])(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": (0, _interopRequireDefault2["default"])(require("./../intl/da-DK.json"))["default"],
  "de-DE": (0, _interopRequireDefault2["default"])(require("./../intl/de-DE.json"))["default"],
  "en-US": (0, _interopRequireDefault2["default"])(require("./../intl/en-US.json"))["default"],
  "es-ES": (0, _interopRequireDefault2["default"])(require("./../intl/es-ES.json"))["default"],
  "fi-FI": (0, _interopRequireDefault2["default"])(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": (0, _interopRequireDefault2["default"])(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": (0, _interopRequireDefault2["default"])(require("./../intl/hu-HU.json"))["default"],
  "it-IT": (0, _interopRequireDefault2["default"])(require("./../intl/it-IT.json"))["default"],
  "ja-JP": (0, _interopRequireDefault2["default"])(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": (0, _interopRequireDefault2["default"])(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": (0, _interopRequireDefault2["default"])(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": (0, _interopRequireDefault2["default"])(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": (0, _interopRequireDefault2["default"])(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": (0, _interopRequireDefault2["default"])(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": (0, _interopRequireDefault2["default"])(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": (0, _interopRequireDefault2["default"])(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": (0, _interopRequireDefault2["default"])(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": (0, _interopRequireDefault2["default"])(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": (0, _interopRequireDefault2["default"])(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": (0, _interopRequireDefault2["default"])(require("./../intl/zh-TW.json"))["default"]
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/inputgroup/index.css");

  require("../../spectrum-css/inputgroup/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/inputgroup/index-lg.css");
} else {
  require("../../spectrum-css/inputgroup/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/inputgroup/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/inputgroup/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/inputgroup/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/inputgroup/multiStops/darkest.css");

var getLabel = function getLabel(o) {
  return typeof o === 'string' ? o : o.label;
};

var formatMessage = (0, _intl.messageFormatter)(intlMessages);

var ComboBox = (0, _convertUnsafeMethod["default"])(_class = (0, _autobindDecorator["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(ComboBox, _React$Component);

  function ComboBox(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, ComboBox);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ComboBox).call(this, props));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      open: false,
      count: null
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "setAutocompleteRef", function (a) {
      return _this.autocomplete = a;
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "setTextFieldRef", function (t) {
      return _this.textfield = t;
    });
    _this.comboBoxId = (0, _createId["default"])();
    return _this;
  }

  (0, _createClass2["default"])(ComboBox, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(props) {
      if (props.showMenu != null && props.showMenu !== this.props.showMenu) {
        this.setState({
          open: props.showMenu
        });
      }
    }
  }, {
    key: "onButtonClick",
    value: function onButtonClick() {
      this.textfield.focus();
      this.autocomplete.toggleMenu();
    }
  }, {
    key: "getCompletions",
    value: function getCompletions(text) {
      if (this.shouldFilter(text)) {
        // show options that start with the text first, then
        // show other options containing the text
        return this.props.options.filter(function (option) {
          return getLabel(option).toLowerCase().includes(text.toLowerCase());
        }).sort(function (a, b) {
          var aStartsWithText = getLabel(a).toLowerCase().startsWith(text.toLowerCase());
          var bStartsWithText = getLabel(b).toLowerCase().startsWith(text.toLowerCase());

          if (aStartsWithText && !bStartsWithText) {
            return -1;
          }

          if (!aStartsWithText && bStartsWithText) {
            return 1;
          }

          return 0;
        });
      }

      return this.props.options;
    }
  }, {
    key: "shouldFilter",
    value: function shouldFilter(text) {
      // if any input has been made since we opened the menu, let's filter
      if (this.changeSinceOpen) {
        return true;
      } // if the current value isn't in the list, let's filter


      return !this.props.options.some(function (o) {
        return getLabel(o) === text;
      });
    }
  }, {
    key: "onMenuShow",
    value: function onMenuShow() {
      if (this.props.showMenu !== null) {
        this.setState({
          open: true
        });
      }
    }
  }, {
    key: "onMenuHide",
    value: function onMenuHide() {
      this.changeSinceOpen = false;

      if (this.props.showMenu !== null) {
        this.setState({
          open: false
        });
      }
    }
  }, {
    key: "onChange",
    value: function onChange(value) {
      this.changeSinceOpen = true;
      var count = value ? this.getCompletions(value).length : null;
      this.setState({
        count: count
      });
    }
  }, {
    key: "getButtonLabel",
    value: function getButtonLabel() {
      var options = this.props.options;
      var count = this.state.count;
      var key = 'Show suggestions';

      if (count === null && options.length > 0) {
        count = options.length;
      }

      if (count === 0) {
        key = 'No matching results';
      } else if (count > 1) {
        key = 'Show {0} suggestions';
      } else if (count === 1) {
        key = 'Show suggestion';
      }

      return formatMessage(key, [count]);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          _this$props$id = _this$props.id,
          id = _this$props$id === void 0 ? this.comboBoxId : _this$props$id,
          className = _this$props.className,
          value = _this$props.value,
          disabled = _this$props.disabled,
          required = _this$props.required,
          invalid = _this$props.invalid,
          quiet = _this$props.quiet,
          onChange = _this$props.onChange,
          onSelect = _this$props.onSelect,
          renderItem = _this$props.renderItem,
          onMenuToggle = _this$props.onMenuToggle,
          showMenu = _this$props.showMenu,
          props = (0, _objectWithoutProperties2["default"])(_this$props, ["id", "className", "value", "disabled", "required", "invalid", "quiet", "onChange", "onSelect", "renderItem", "onMenuToggle", "showMenu"]);
      return _react["default"].createElement(_Autocomplete["default"], {
        className: (0, _classnames["default"])('spectrum-InputGroup', {
          'spectrum-InputGroup--quiet': quiet
        }, className),
        ref: this.setAutocompleteRef,
        getCompletions: this.getCompletions,
        value: value,
        onChange: (0, _events.chain)(onChange, this.onChange),
        onSelect: onSelect,
        onMenuShow: this.onMenuShow,
        onMenuHide: this.onMenuHide,
        showMenu: showMenu,
        onMenuToggle: onMenuToggle,
        renderItem: renderItem
      }, _react["default"].createElement(_Textfield["default"], (0, _extends2["default"])({
        className: (0, _classnames["default"])('spectrum-InputGroup-field')
      }, props, {
        id: id,
        ref: this.setTextFieldRef,
        disabled: disabled,
        required: required,
        invalid: invalid,
        autocompleteInput: true,
        quiet: quiet
      })), _react["default"].createElement(_Button["default"], {
        id: "".concat(id, "-button"),
        type: "button",
        variant: "field",
        onClick: this.onButtonClick,
        onMouseDown: function onMouseDown(e) {
          return e.preventDefault();
        },
        onMouseUp: function onMouseUp(e) {
          return e.preventDefault();
        },
        disabled: disabled,
        required: required,
        invalid: invalid,
        quiet: quiet,
        selected: showMenu !== undefined ? showMenu : this.state.open,
        "aria-label": this.getButtonLabel(),
        "aria-labelledby": "".concat(id, " ").concat(id, "-button"),
        tabIndex: "-1"
      }, _react["default"].createElement(_ChevronDownMedium["default"], {
        size: null,
        className: "spectrum-InputGroup-icon"
      })));
    }
  }]);
  return ComboBox;
}(_react["default"].Component), (0, _defineProperty2["default"])(_class2, "propTypes", {
  /**
   * Value of the input
   */
  value: _propTypes["default"].string,

  /**
   * List of options for the combobox menu
   */
  options: _propTypes["default"].array,

  /**
   * Greys out the control
   */
  disabled: _propTypes["default"].bool,

  /**
   * Marks it as required
   */
  required: _propTypes["default"].bool,

  /**
   * Marks the input as invalid
   */
  invalid: _propTypes["default"].bool,

  /**
   * Variant that changes the look, removes some borders
   */
  quiet: _propTypes["default"].bool,

  /**
   * A callback for both show and hide, event is false if hiding, true if showing.
   * Reason for this was to add a controlled state in a backwards compatible way,
   * we couldn't use show/hide props for that, so we needed a new one.
   */
  onMenuToggle: _propTypes["default"].func,

  /**
   * Controlled state for showing/hiding menu.
   */
  showMenu: _propTypes["default"].bool,

  /**
   * A function that returns a wrapper component to render a list item label.
   * Useful in providing custom html to the rendered label.
   */
  renderItem: _propTypes["default"].func
}), (0, _defineProperty2["default"])(_class2, "defaultProps", {
  options: [],
  disabled: false,
  required: false,
  invalid: false,
  quiet: false
}), _temp)) || _class) || _class;

exports["default"] = ComboBox;
