// @react/react-spectrum/Breadcrumbs/js/Breadcrumbs.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _objectWithoutProperties2 = _interopRequireDefault3(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault3(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault3(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault3(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault3(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault3(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault3(require("@babel/runtime/helpers/defineProperty"));

var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));

var _ChevronRightSmall = _interopRequireDefault3(require("../../Icon/core/ChevronRightSmall"));

var _classnames = _interopRequireDefault3(require("classnames"));

var _icon = require("../../utils/icon");

var _focusRing = require("../../utils/focusRing");

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault3(require("prop-types"));

var _react = _interopRequireDefault3(require("react"));

require("../style/index.css");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var intlMessages = {
  "cs-CZ": (0, _interopRequireDefault2["default"])(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": (0, _interopRequireDefault2["default"])(require("./../intl/da-DK.json"))["default"],
  "de-DE": (0, _interopRequireDefault2["default"])(require("./../intl/de-DE.json"))["default"],
  "en-US": (0, _interopRequireDefault2["default"])(require("./../intl/en-US.json"))["default"],
  "es-ES": (0, _interopRequireDefault2["default"])(require("./../intl/es-ES.json"))["default"],
  "fi-FI": (0, _interopRequireDefault2["default"])(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": (0, _interopRequireDefault2["default"])(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": (0, _interopRequireDefault2["default"])(require("./../intl/hu-HU.json"))["default"],
  "it-IT": (0, _interopRequireDefault2["default"])(require("./../intl/it-IT.json"))["default"],
  "ja-JP": (0, _interopRequireDefault2["default"])(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": (0, _interopRequireDefault2["default"])(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": (0, _interopRequireDefault2["default"])(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": (0, _interopRequireDefault2["default"])(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": (0, _interopRequireDefault2["default"])(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": (0, _interopRequireDefault2["default"])(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": (0, _interopRequireDefault2["default"])(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": (0, _interopRequireDefault2["default"])(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": (0, _interopRequireDefault2["default"])(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": (0, _interopRequireDefault2["default"])(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": (0, _interopRequireDefault2["default"])(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": (0, _interopRequireDefault2["default"])(require("./../intl/zh-TW.json"))["default"]
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/breadcrumb/index.css");

  require("../../spectrum-css/breadcrumb/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/breadcrumb/index-lg.css");
} else {
  require("../../spectrum-css/breadcrumb/index.css");
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require("../../spectrum-css/breadcrumb/multiStops/light.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require("../../spectrum-css/breadcrumb/multiStops/dark.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require("../../spectrum-css/breadcrumb/multiStops/lightest.css");
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require("../../spectrum-css/breadcrumb/multiStops/darkest.css");
var formatMessage = (0, _intl.messageFormatter)(intlMessages);

var Breadcrumbs =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(Breadcrumbs, _React$Component);

  function Breadcrumbs() {
    (0, _classCallCheck2["default"])(this, Breadcrumbs);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Breadcrumbs).apply(this, arguments));
  }

  (0, _createClass2["default"])(Breadcrumbs, [{
    key: "handleFocus",
    value: function handleFocus(e) {
      e.target.classList.add(_focusRing.FOCUS_RING_CLASSNAME);
    }
  }, {
    key: "handleBlur",
    value: function handleBlur(e) {
      e.target.classList.remove(_focusRing.FOCUS_RING_CLASSNAME);
    }
  }, {
    key: "renderLink",
    value: function renderLink(item, i) {
      var _this$props = this.props,
          items = _this$props.items,
          renderLink = _this$props.renderLink;
      var isHref = !!item.href;
      var isCurrent = this.isCurrent(i);
      var hasItems = items.length > 1;
      var linkProps = {
        className: 'spectrum-Breadcrumbs-itemLink',
        role: !isHref ? 'link' : null,
        href: !isCurrent ? item.href : null,
        target: !isCurrent && isHref ? '_self' : null,
        onClick: hasItems && !isCurrent ? this.props.onBreadcrumbClick.bind(null, item, items.length - i - 1) : undefined,
        onFocus: hasItems && !isCurrent ? this.handleFocus.bind(this) : null,
        onBlur: hasItems && !isCurrent ? this.handleBlur.bind(this) : null,
        'aria-current': isCurrent ? 'page' : null,
        tabIndex: !isHref && !isCurrent ? 0 : null
      };

      if (renderLink) {
        return renderLink(_objectSpread({}, linkProps, {
          label: item.label
        }));
      }

      var Element = !isHref ? 'div' : 'a';
      return _react["default"].createElement(Element, linkProps, item.label);
    }
  }, {
    key: "isCurrent",
    value: function isCurrent(i) {
      return i === this.props.items.length - 1;
    }
  }, {
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props2 = this.props,
          items = _this$props2.items,
          icon = _this$props2.icon,
          className = _this$props2.className,
          variant = _this$props2.variant,
          ariaLevel = _this$props2.ariaLevel,
          id = _this$props2.id,
          otherProps = (0, _objectWithoutProperties2["default"])(_this$props2, ["items", "icon", "className", "variant", "ariaLevel", "id"]);
      var isTitleVariant = variant === Breadcrumbs.variant.TITLE;

      if (!otherProps['aria-label']) {
        otherProps['aria-label'] = formatMessage('Breadcrumbs');
      }

      return _react["default"].createElement("nav", {
        id: id,
        "aria-label": otherProps['aria-label'],
        "aria-labelledby": otherProps['aria-labelledby']
      }, (0, _icon.cloneIcon)(icon, {
        size: 'S',
        className: 'react-spectrum-Breadcrumbs-icon'
      }), _react["default"].createElement("ul", {
        className: (0, _classnames["default"])('spectrum-Breadcrumbs', {
          'spectrum-Breadcrumbs--title': isTitleVariant
        }, className)
      }, items.map(function (item, i) {
        return _react["default"].createElement("li", {
          key: "spectrum-Breadcrumb-".concat(i),
          className: "spectrum-Breadcrumbs-item"
        }, _this.isCurrent(i) && isTitleVariant ? _react["default"].createElement("h1", {
          className: "spectrum-Heading--pageTitle",
          "aria-level": ariaLevel
        }, _this.renderLink(item, i)) : _this.renderLink(item, i), !_this.isCurrent(i) && _react["default"].createElement(_ChevronRightSmall["default"], {
          size: null,
          className: "spectrum-Breadcrumbs-itemSeparator"
        }));
      })));
    }
  }]);
  return Breadcrumbs;
}(_react["default"].Component);

exports["default"] = Breadcrumbs;
(0, _defineProperty2["default"])(Breadcrumbs, "variant", {
  NONE: 'none',
  TITLE: 'title'
});
(0, _defineProperty2["default"])(Breadcrumbs, "propTypes", {
  /**
   * Array of item objects included in list of breadcrumbs.
   * Each item must have a label string property, which will serve as the rendered text for the breadcrumb.
   * Each item may also have an href string property, which should be a valid URL to open the breadcrumb location.
   */
  items: _propTypes["default"].arrayOf(_propTypes["default"].oneOfType([_propTypes["default"].shape({
    href: _propTypes["default"].string,
    label: _propTypes["default"].string.isRequired
  }), _propTypes["default"].shape({
    label: _propTypes["default"].string.isRequired
  })])),

  /**
   * For best results, use a React Spectrum Icon
   */
  icon: _propTypes["default"].node,

  /**
   * Will set the last breadcrumb on a new line and give it a <h> tag
   */
  variant: _propTypes["default"].oneOf([Breadcrumbs.variant.NONE, Breadcrumbs.variant.TITLE]),

  /**
   * Will not change the appearance
   */
  ariaLevel: _propTypes["default"].number,

  /**
   * Called when a breadcrumb is clicked with an object containing the label of the clicked breadcrumb
   * @callback Breadcrumbs~onBreadcrumbClick
   * @param {Object} event - Event object
   * @param {string} event.label - label of breadcrumb clicked
   */
  onBreadcrumbClick: _propTypes["default"].func,

  /**
   * A function that returns a href wrapper component.
   * Useful in providing custom href component(eg. Link from react-router-dom).
   *
   * ```js
   *  <Breadcrumbs renderLink={(props) => <Link {...props} to="/">Foo</Link>} items={items} />
   * ```
   */
  renderLink: _propTypes["default"].func
});
(0, _defineProperty2["default"])(Breadcrumbs, "defaultProps", {
  items: [],
  icon: null,
  variant: Breadcrumbs.variant.NONE,
  ariaLevel: null,
  onBreadcrumbClick: function onBreadcrumbClick() {}
});
