// @react/react-spectrum/SearchWithin/js/SearchWithin.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault3(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault3(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault3(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault3(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault3(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault3(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault3(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault3(require("@babel/runtime/helpers/defineProperty"));

var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));

var _classnames = _interopRequireDefault3(require("classnames"));

var _createId = _interopRequireDefault3(require("../../utils/createId"));

var _filterDOMProps = _interopRequireDefault3(require("../../utils/filterDOMProps"));

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault3(require("prop-types"));

var _react = _interopRequireDefault3(require("react"));

var _Search = _interopRequireDefault3(require("../../Search"));

var _Select = _interopRequireDefault3(require("../../Select"));

/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2019 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.
**************************************************************************/
var intlMessages = {
  "cs-CZ": (0, _interopRequireDefault2["default"])(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": (0, _interopRequireDefault2["default"])(require("./../intl/da-DK.json"))["default"],
  "de-DE": (0, _interopRequireDefault2["default"])(require("./../intl/de-DE.json"))["default"],
  "en-US": (0, _interopRequireDefault2["default"])(require("./../intl/en-US.json"))["default"],
  "es-ES": (0, _interopRequireDefault2["default"])(require("./../intl/es-ES.json"))["default"],
  "fi-FI": (0, _interopRequireDefault2["default"])(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": (0, _interopRequireDefault2["default"])(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": (0, _interopRequireDefault2["default"])(require("./../intl/hu-HU.json"))["default"],
  "it-IT": (0, _interopRequireDefault2["default"])(require("./../intl/it-IT.json"))["default"],
  "ja-JP": (0, _interopRequireDefault2["default"])(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": (0, _interopRequireDefault2["default"])(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": (0, _interopRequireDefault2["default"])(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": (0, _interopRequireDefault2["default"])(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": (0, _interopRequireDefault2["default"])(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": (0, _interopRequireDefault2["default"])(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": (0, _interopRequireDefault2["default"])(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": (0, _interopRequireDefault2["default"])(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": (0, _interopRequireDefault2["default"])(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": (0, _interopRequireDefault2["default"])(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": (0, _interopRequireDefault2["default"])(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": (0, _interopRequireDefault2["default"])(require("./../intl/zh-TW.json"))["default"]
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require("../../spectrum-css/searchwithin/index.css");

  require("../../spectrum-css/searchwithin/index-diff.css");
} else if (process.env.SCALE_LARGE) {
  require("../../spectrum-css/searchwithin/index-lg.css");
} else {
  require("../../spectrum-css/searchwithin/index.css");
}

require('../style/index.css');

var formatMessage = (0, _intl.messageFormatter)(intlMessages);
/**
 * A SearchWithin component displays a dropdown next to a search input, typically
 * used to select the scope of a search.
 */

var SearchWithin =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(SearchWithin, _React$Component);

  function SearchWithin(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, SearchWithin);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(SearchWithin).call(this, props));
    _this.outerId = (0, _createId["default"])();
    return _this;
  }

  (0, _createClass2["default"])(SearchWithin, [{
    key: "getChildId",
    value: function getChildId(childName) {
      return this.outerId + '-' + childName;
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          scope = _this$props.scope,
          defaultScope = _this$props.defaultScope,
          value = _this$props.value,
          defaultValue = _this$props.defaultValue,
          onScopeChange = _this$props.onScopeChange,
          onValueChange = _this$props.onValueChange,
          onSubmit = _this$props.onSubmit,
          _this$props$disabled = _this$props.disabled,
          disabled = _this$props$disabled === void 0 ? false : _this$props$disabled,
          _this$props$placehold = _this$props.placeholder,
          placeholder = _this$props$placehold === void 0 ? '' : _this$props$placehold,
          _this$props$id = _this$props.id,
          id = _this$props$id === void 0 ? this.getChildId('search') : _this$props$id,
          _this$props$selectId = _this$props.selectId,
          selectId = _this$props$selectId === void 0 ? this.getChildId('select') : _this$props$selectId,
          ariaLabelledby = _this$props['aria-labelledby'],
          _this$props$ariaLabe = _this$props['aria-label'],
          ariaLabel = _this$props$ariaLabe === void 0 ? !ariaLabelledby ? formatMessage('Search within') : null : _this$props$ariaLabe,
          autoFocus = _this$props.autoFocus,
          className = _this$props.className,
          scopeOptions = _this$props.scopeOptions,
          otherProps = (0, _objectWithoutProperties2["default"])(_this$props, ["scope", "defaultScope", "value", "defaultValue", "onScopeChange", "onValueChange", "onSubmit", "disabled", "placeholder", "id", "selectId", "aria-labelledby", "aria-label", "autoFocus", "className", "scopeOptions"]);
      var formattedScopeOptions = scopeOptions.map(function (scope) {
        return typeof scope === 'string' ? {
          label: scope,
          value: scope
        } : scope;
      });

      if (ariaLabelledby) {
        if (ariaLabel) {
          ariaLabelledby += " ".concat(this.outerId);
        }
      } else {
        ariaLabelledby = this.outerId;
      }

      var selectProps = {};

      if (scope) {
        selectProps.value = scope;
      } else {
        selectProps.defaultValue = defaultScope ? defaultScope : formattedScopeOptions[0].value;
      }

      var select = _react["default"].createElement(_Select["default"], (0, _extends2["default"])({
        id: selectId,
        "aria-labelledby": ariaLabelledby,
        onChange: onScopeChange,
        options: formattedScopeOptions,
        disabled: disabled,
        required: true,
        flexible: true
      }, selectProps));

      var searchProps = {}; // comparison is on purpose, we don't want to mix undefined with null in equality

      if (value != null) {
        searchProps.value = value;
      } else if (defaultValue) {
        searchProps.defaultValue = defaultValue;
      }

      var search = _react["default"].createElement(_Search["default"], (0, _extends2["default"])({
        id: id,
        "aria-labelledby": ariaLabelledby + " ".concat(selectId, "-value"),
        placeholder: placeholder,
        onChange: onValueChange,
        onSubmit: onSubmit,
        disabled: disabled,
        role: "presentation",
        autoFocus: autoFocus
      }, searchProps));

      return _react["default"].createElement("div", (0, _extends2["default"])({
        className: (0, _classnames["default"])('spectrum-SearchWithin', 'react-spectrum-SearchWithin', className),
        "aria-label": ariaLabel,
        "aria-labelledby": ariaLabelledby,
        id: this.outerId,
        role: "search"
      }, (0, _filterDOMProps["default"])(otherProps)), select, search);
    }
  }]);
  return SearchWithin;
}(_react["default"].Component);

exports["default"] = SearchWithin;
(0, _defineProperty2["default"])(SearchWithin, "propTypes", {
  /**
   * Scope options to be shown in the dropdown
   */
  scopeOptions: _propTypes["default"].arrayOf(_propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].shape({
    label: _propTypes["default"].string,
    value: _propTypes["default"].string
  })])).isRequired,

  /**
   * The currently selected scope value (controlled).
   */
  scope: _propTypes["default"].string,

  /**
   * The default selected scope value (uncontrolled).
   */
  defaultScope: _propTypes["default"].string,

  /**
   * The current value of the textfield (controlled).
   */
  value: _propTypes["default"].string,

  /**
   * The default value of the textfield (uncontrolled).
   */
  defaultValue: _propTypes["default"].string,

  /**
   * A placeholder for the textfield.
   */
  placeholder: _propTypes["default"].string,

  /**
   * Whether the field is disabled
   */
  disabled: _propTypes["default"].bool,

  /**
   * A callback for when the field is submitted
   */
  onSubmit: _propTypes["default"].func,

  /**
   * A callback for when the textfield changes.
   */
  onValueChange: _propTypes["default"].func,

  /**
   * A callback for when the scope changes
   */
  onScopeChange: _propTypes["default"].func,

  /**
  * Class given to SearchWithin
  */
  className: _propTypes["default"].string
});
