// @react/react-spectrum/Shell/js/ShellSolutionSwitcher.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ShellSolutionSwitcher;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _Apps = _interopRequireDefault(require("../../Icon/Apps"));

var _Button = _interopRequireDefault(require("../../Button"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _ShellMenu = _interopRequireDefault(require("./ShellMenu"));

require("../style/ShellSolutionSwitcher.css");

/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2019 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.
**************************************************************************/
function ShellSolutionSwitcher(_ref) {
  var className = _ref.className,
      children = _ref.children,
      otherProps = (0, _objectWithoutProperties2["default"])(_ref, ["className", "children"]);
  return _react["default"].createElement(_ShellMenu["default"], (0, _extends2["default"])({
    placement: "top",
    animateFrom: "top",
    full: true,
    dark: true,
    top: true,
    target: _react["default"].createElement(_Button["default"], {
      className: "coral3-Shell-menu-button",
      variant: "minimal",
      icon: _react["default"].createElement(_Apps["default"], null)
    })
  }, otherProps), _react["default"].createElement("div", {
    className: (0, _classnames["default"])('coral3-Shell-solutionSwitcher', className)
  }, _react["default"].createElement("div", {
    className: "coral3-Shell-solutionSwitcher-content"
  }, children)));
}

ShellSolutionSwitcher.displayName = 'ShellSolutionSwitcher';
