// @react/react-spectrum/Shell/js/ShellHelp.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _Button = _interopRequireDefault(require("../../Button"));

var _classnames = _interopRequireDefault(require("classnames"));

var _HelpOutline = _interopRequireDefault(require("../../Icon/HelpOutline"));

var _Link = _interopRequireDefault(require("../../Link"));

var _List = require("../../List");

var _react = _interopRequireWildcard(require("react"));

var _Search = _interopRequireDefault(require("../../Search"));

var _ShellMenu = _interopRequireDefault(require("./ShellMenu"));

var _Wait = _interopRequireDefault(require("../../Wait"));

require("../style/ShellHelp.css");

/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2019 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.
**************************************************************************/
var ShellHelp =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(ShellHelp, _Component);

  function ShellHelp() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, ShellHelp);

    for (var _len = arguments.length, _args = new Array(_len), _key = 0; _key < _len; _key++) {
      _args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(ShellHelp)).call.apply(_getPrototypeOf2, [this].concat(_args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      searchTerm: ''
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleSearch", function (val) {
      var onSearch = _this.props.onSearch;
      onSearch(val);

      _this.setState({
        searchTerm: val
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleVisible", function () {
      if (_this.contentRef) {
        _this.contentRef.querySelector('.spectrum-Search-input').focus();
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleResultClick", function () {
      if (_this.props.onResultClick) {
        var _this$props;

        (_this$props = _this.props).onResultClick.apply(_this$props, arguments);
      }
    });
    return _this;
  }

  (0, _createClass2["default"])(ShellHelp, [{
    key: "renderResults",
    value: function renderResults() {
      var searchResults = this.props.searchResults;

      if (searchResults) {
        return this.renderSearchResults();
      }

      return this.renderDefaultResults();
    }
  }, {
    key: "renderSearchResults",
    value: function renderSearchResults() {
      var _this2 = this;

      var _this$props2 = this.props,
          searchResults = _this$props2.searchResults,
          numTotalResults = _this$props2.numTotalResults,
          moreSearchResultsUrl = _this$props2.moreSearchResultsUrl;
      var searchTerm = this.state.searchTerm; // https://github.com/adobe/react-coral/issues/134
      // unicode for: '&nbsp; &bull; &nbsp';

      var separator = "\xA0 \u2022 \xA0";
      var allResults = {
        href: "".concat(moreSearchResultsUrl, "?q=").concat(searchTerm),
        title: "See all ".concat(numTotalResults, " results")
      };
      return _react["default"].createElement(_List.List, {
        className: "coral3-Shell-help-results"
      }, numTotalResults !== 0 && searchResults && searchResults.length ? searchResults.map(function (result, index) {
        return _react["default"].createElement(_List.ListItem, {
          key: index,
          onClick: _this2.handleResultClick.bind(_this2, result)
        }, _react["default"].createElement(_Link["default"], {
          href: result.href,
          target: "_blank",
          rel: "noopener noreferrer"
        }, _react["default"].createElement("span", null, result.title, _react["default"].createElement("div", {
          className: "coral3-Shell-help-result-description"
        }, Array.isArray(result.tags) ? result.tags.join(separator) : ''))));
      }).concat(_react["default"].createElement(_List.ListItem, {
        key: "all-results",
        onClick: this.handleResultClick.bind(this, allResults)
      }, _react["default"].createElement(_Link["default"], {
        className: "spectrum-Link coral3-Shell-help-allResults",
        href: allResults.href,
        target: "_blank",
        rel: "noopener noreferrer"
      }, allResults.title))) : _react["default"].createElement("div", {
        className: "coral3-Shell-help-resultMessage"
      }, _react["default"].createElement("div", {
        className: "coral3-Shell-help-resultMessage-container"
      }, _react["default"].createElement("div", {
        className: "coral-Heading--1 coral3-Shell-help-resultMessage-heading"
      }, "No results found"))));
    }
  }, {
    key: "renderDefaultResults",
    value: function renderDefaultResults() {
      var _this3 = this;

      var defaultResults = this.props.defaultResults;
      return _react["default"].createElement(_List.List, {
        className: "coral3-Shell-help-items"
      }, defaultResults && defaultResults.length ? defaultResults.map(function (result, index) {
        return _react["default"].createElement(_List.ListItem, {
          icon: result.icon,
          key: index,
          onClick: _this3.handleResultClick.bind(_this3, result)
        }, _react["default"].createElement(_Link["default"], {
          className: "coral3-Shell-help-item",
          href: result.href,
          target: "_blank",
          rel: "noopener noreferrer"
        }, result.label));
      }) : null);
    }
  }, {
    key: "render",
    value: function render() {
      var _this4 = this;

      var _this$props3 = this.props,
          className = _this$props3.className,
          loading = _this$props3.loading,
          otherProps = (0, _objectWithoutProperties2["default"])(_this$props3, ["className", "loading"]);
      return _react["default"].createElement(_ShellMenu["default"], (0, _extends2["default"])({
        placement: "right",
        animateFrom: "top",
        target: _react["default"].createElement(_Button["default"], {
          className: "coral3-Shell-menu-button",
          variant: "minimal",
          icon: _react["default"].createElement(_HelpOutline["default"], null)
        }),
        onVisible: this.handleVisible,
        onHidden: this.props.onHide
      }, otherProps), _react["default"].createElement("div", {
        ref: function ref(el) {
          _this4.contentRef = el;
        },
        className: (0, _classnames["default"])('coral3-Shell-help', className)
      }, _react["default"].createElement(_Search["default"], {
        className: "coral3-Shell-help-search",
        placeholder: "Search Help",
        onSubmit: this.handleSearch,
        onChange: this.props.onChange,
        quiet: true
      }), loading ? _react["default"].createElement("div", {
        className: "coral3-Shell-help-loading"
      }, _react["default"].createElement(_Wait["default"], {
        className: "coral3-Shell-help-loading-wait"
      }), _react["default"].createElement("span", {
        className: "coral-Heading--2 coral3-Shell-help-loading-info"
      }, "Searching Help...")) : this.renderResults()));
    }
  }]);
  return ShellHelp;
}(_react.Component);

exports["default"] = ShellHelp;
(0, _defineProperty2["default"])(ShellHelp, "defaultProps", {
  defaultResults: [],
  onSearch: function onSearch() {},
  onChange: function onChange() {},
  onHide: function onHide() {},
  onResultClick: function onResultClick() {}
});
ShellHelp.displayName = 'ShellHelp';
