// @react/react-spectrum/Shell/js/ShellActions.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireWildcard(require("react"));

/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2019 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.
**************************************************************************/
var ShellHeaderActions =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(ShellHeaderActions, _Component);

  function ShellHeaderActions() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, ShellHeaderActions);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(ShellHeaderActions)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      openIndex: -1
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleMenuOpen", function (e, index) {
      _this.setState({
        openIndex: index
      });
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleMenuClose", function () {
      _this.setState({
        openIndex: -1
      });
    });
    return _this;
  }

  (0, _createClass2["default"])(ShellHeaderActions, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          className = _this$props.className,
          children = _this$props.children,
          betaFeedback = _this$props.betaFeedback,
          otherProps = (0, _objectWithoutProperties2["default"])(_this$props, ["className", "children", "betaFeedback"]);
      var openIndex = this.state.openIndex;
      var index = 0;
      return _react["default"].createElement("div", (0, _extends2["default"])({
        className: (0, _classnames["default"])('coral3-Shell-header-actions', className)
      }, otherProps), betaFeedback, _react["default"].createElement("div", {
        className: "coral3-Shell-menubar"
      }, children && _react["default"].Children.map(children, function (child) {
        if ((0, _typeof2["default"])(child) === 'object' && child && child.type) {
          // Is this a react element?
          var childProps = child.type.name === 'Button' ? {} : {
            onOpen: _this2.handleMenuOpen,
            onClose: _this2.handleMenuClose,
            defaultOpen: index === openIndex,
            index: index++
          };
          return _react["default"].cloneElement(child, childProps);
        } // Otherwise, the child is a string


        return child;
      })));
    }
  }]);
  return ShellHeaderActions;
}(_react.Component);

exports["default"] = ShellHeaderActions;
ShellHeaderActions.displayName = 'ShellHeaderActions';
