// @react/react-spectrum/Shell/js/ShellSolutionGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2019 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.
**************************************************************************/
var ShellSolutionGroup =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(ShellSolutionGroup, _Component);

  function ShellSolutionGroup() {
    (0, _classCallCheck2["default"])(this, ShellSolutionGroup);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ShellSolutionGroup).apply(this, arguments));
  }

  (0, _createClass2["default"])(ShellSolutionGroup, [{
    key: "renderPrimarySolutions",
    value: function renderPrimarySolutions(children) {
      var entitledChildren = children.filter(function (child) {
        return child.props && child.props.entitled;
      });
      var unentitledChildren = children.filter(function (child) {
        return child.props && child.props.entitled == null;
      });
      return _react["default"].createElement("div", {
        className: "coral-Shell-solutions-container"
      }, _react["default"].createElement("div", {
        className: "coral-Shell-solutions"
      }, entitledChildren), _react["default"].createElement("div", {
        className: "coral-Shell-solutions"
      }, unentitledChildren));
    }
  }, {
    key: "renderSecondarySolutions",
    value: function renderSecondarySolutions(children) {
      return _react["default"].createElement("div", {
        className: "coral-Shell-solutions coral-Shell-solutions--secondary"
      }, _react["default"].Children.map(children, function (child) {
        return _react["default"].cloneElement(child, {
          target: '_blank'
        });
      }));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          secondary = _this$props.secondary,
          children = _this$props.children;

      if (secondary) {
        return this.renderSecondarySolutions(children);
      }

      return this.renderPrimarySolutions(children);
    }
  }]);
  return ShellSolutionGroup;
}(_react.Component);

exports["default"] = ShellSolutionGroup;
(0, _defineProperty2["default"])(ShellSolutionGroup, "defaultProps", {
  secondary: false
});
ShellSolutionGroup.displayName = 'ShellSolutionGroup';
