// @react/react-spectrum/Shell/js/ShellOrgSwitcher.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _Button = _interopRequireDefault(require("../../Button"));

var _classnames = _interopRequireDefault(require("classnames"));

var _convertUnsafeMethod = _interopRequireDefault(require("../../utils/convertUnsafeMethod"));

var _react = _interopRequireWildcard(require("react"));

var _Search = _interopRequireDefault(require("../../Search"));

var _SelectList = _interopRequireDefault(require("../../SelectList"));

var _ShellMenu = _interopRequireDefault(require("./ShellMenu"));

require("../style/ShellOrgSwitcher.css");

var _class, _class2, _temp;

var ShellOrgSwitcher = (0, _convertUnsafeMethod["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(ShellOrgSwitcher, _Component);

  function ShellOrgSwitcher() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, ShellOrgSwitcher);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(ShellOrgSwitcher)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "state", {
      searchTerm: '',
      visibleOptions: _this.filterVisibleOptions(_this.props.options, '')
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleVisible", function () {
      if (_this.contentRef) {
        _this.contentRef.querySelector('.spectrum-Search-input').focus();
      }
    });
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "handleSearchChange", function (searchTerm) {
      var options = _this.props.options;

      _this.setState({
        searchTerm: searchTerm,
        visibleOptions: _this.filterVisibleOptions(options, searchTerm)
      });
    });
    return _this;
  }

  (0, _createClass2["default"])(ShellOrgSwitcher, [{
    key: "UNSAFE_componentWillReceiveProps",
    value: function UNSAFE_componentWillReceiveProps(nextProps) {
      if (nextProps.options !== this.props.options) {
        this.setState({
          visibleOptions: this.filterVisibleOptions(nextProps.options, this.state.searchTerm)
        });
      }
    }
  }, {
    key: "filterVisibleOptions",
    value: function filterVisibleOptions(options, searchTerm) {
      return options.filter(function (option) {
        return option.label.toLowerCase().indexOf(searchTerm) !== -1;
      });
    }
  }, {
    key: "getSelectedLabel",
    value: function getSelectedLabel() {
      var _this$props = this.props,
          options = _this$props.options,
          value = _this$props.value;
      var selectedOptions = options.filter(function (option) {
        return option.value === value;
      });

      if (selectedOptions.length) {
        return selectedOptions[0].label;
      }

      return value || '';
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props2 = this.props,
          value = _this$props2.value,
          className = _this$props2.className,
          manageOrgsUrl = _this$props2.manageOrgsUrl,
          onOrgChange = _this$props2.onOrgChange,
          otherProps = (0, _objectWithoutProperties2["default"])(_this$props2, ["value", "className", "manageOrgsUrl", "onOrgChange"]);
      var visibleOptions = this.state.visibleOptions;
      return _react["default"].createElement(_ShellMenu["default"], (0, _extends2["default"])({
        placement: "right",
        animateFrom: "top",
        target: _react["default"].createElement(_Button["default"], {
          variant: "minimal",
          className: "coral3-Shell-menu-button"
        }, this.getSelectedLabel()),
        onVisible: this.handleVisible
      }, otherProps), _react["default"].createElement("div", {
        ref: function ref(el) {
          _this2.contentRef = el;
        },
        className: (0, _classnames["default"])('coral3-Shell-orgSwitcher', className)
      }, _react["default"].createElement(_Search["default"], {
        className: "coral3-Shell-orgSwitcher-search",
        placeholder: "Search Organizations",
        onChange: this.handleSearchChange,
        quiet: true
      }), visibleOptions.length !== 0 && _react["default"].createElement(_SelectList["default"], {
        className: "coral3-Shell-orgSwitcher-items",
        value: value,
        options: visibleOptions,
        onChange: onOrgChange
      }), visibleOptions.length === 0 && _react["default"].createElement("div", {
        className: "coral3-Shell-orgSwitcher-resultMessage"
      }, _react["default"].createElement("div", {
        className: "coral3-Shell-orgSwitcher-resultMessage-container"
      }, _react["default"].createElement("div", {
        className: "coral-Heading--1 coral3-Shell-orgSwitcher-resultMessage-heading"
      }, "No organizations found."))), _react["default"].createElement("div", {
        className: "coral3-Shell-orgSwitcher-footer"
      }, _react["default"].createElement(_Button["default"], {
        element: "a",
        variant: "minimal",
        href: manageOrgsUrl,
        block: true
      }, "Manage Organizations"))));
    }
  }]);
  return ShellOrgSwitcher;
}(_react.Component), (0, _defineProperty2["default"])(_class2, "defaultProps", {
  value: '',
  manageOrgsUrl: '#',
  options: [],
  onOrgChange: function onOrgChange() {}
}), _temp)) || _class;

exports["default"] = ShellOrgSwitcher;
ShellOrgSwitcher.displayName = 'ShellOrgSwitcher';
