// @react/react-spectrum/Shell/js/ShellUserProfile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ShellUserProfile;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _Button = _interopRequireDefault(require("../../Button"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _ShellMenu = _interopRequireDefault(require("./ShellMenu"));

require("../style/ShellUserProfile.css");

/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2019 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.
**************************************************************************/
function ShellUserProfile(_ref) {
  var name = _ref.name,
      heading = _ref.heading,
      subheading = _ref.subheading,
      _ref$avatarUrl = _ref.avatarUrl,
      avatarUrl = _ref$avatarUrl === void 0 ? '#' : _ref$avatarUrl,
      _ref$profileUrl = _ref.profileUrl,
      profileUrl = _ref$profileUrl === void 0 ? '#' : _ref$profileUrl,
      _ref$onSignOut = _ref.onSignOut,
      onSignOut = _ref$onSignOut === void 0 ? function () {} : _ref$onSignOut,
      children = _ref.children,
      className = _ref.className,
      otherProps = (0, _objectWithoutProperties2["default"])(_ref, ["name", "heading", "subheading", "avatarUrl", "profileUrl", "onSignOut", "children", "className"]);
  return _react["default"].createElement(_ShellMenu["default"], (0, _extends2["default"])({
    placement: "right",
    animateFrom: "top",
    target: _react["default"].createElement(_Button["default"], {
      className: "coral3-Shell-menu-button",
      variant: "minimal"
    }, _react["default"].createElement("img", {
      src: avatarUrl,
      alt: "",
      "aria-hidden": "true"
    }))
  }, otherProps), _react["default"].createElement("div", {
    className: (0, _classnames["default"])('coral3-Shell-user', className)
  }, _react["default"].createElement("div", {
    className: "coral3-Shell-user-container"
  }, _react["default"].createElement("div", {
    className: "coral3-Shell-user-image"
  }, _react["default"].createElement("img", {
    src: avatarUrl,
    alt: "",
    "aria-hidden": "true",
    className: "coral3-Shell-user-avatar"
  })), _react["default"].createElement("div", {
    className: "coral3-Shell-user-name"
  }, name), _react["default"].createElement("div", {
    className: "coral3-Shell-user-heading"
  }, heading && heading !== 'null' ? heading : ''), _react["default"].createElement("div", {
    className: "coral3-Shell-user-subheading"
  }, subheading && subheading !== 'null' ? subheading : '')), _react["default"].createElement("div", {
    className: "coral3-Shell-user-content"
  }, children), _react["default"].createElement("div", {
    className: "coral3-Shell-user-footer"
  }, _react["default"].createElement(_Button["default"], {
    element: "a",
    variant: "minimal",
    href: profileUrl
  }, "Edit Profile"), _react["default"].createElement(_Button["default"], {
    element: "a",
    variant: "minimal",
    onClick: onSignOut
  }, "Sign Out"))));
}

ShellUserProfile.displayName = 'ShellUserProfile';
