// @react/react-spectrum/Tour/js/Tour.js
"use strict";

var _interopRequireDefault3 = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.STATUS_COMPLETE = exports.STATUS_CANCEL = void 0;

var _classCallCheck2 = _interopRequireDefault3(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault3(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault3(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault3(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault3(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault3(require("@babel/runtime/helpers/defineProperty"));

var _interopRequireDefault2 = _interopRequireDefault3(require("@babel/runtime/helpers/interopRequireDefault"));

var _autobindDecorator = _interopRequireDefault3(require("autobind-decorator"));

var _events = require("../../utils/events");

var _intl = require("../../utils/intl");

var _propTypes = _interopRequireDefault3(require("prop-types"));

var _react = _interopRequireDefault3(require("react"));

var _class, _class2, _temp;

var intlMessages = {
  "cs-CZ": (0, _interopRequireDefault2["default"])(require("./../intl/cs-CZ.json"))["default"],
  "da-DK": (0, _interopRequireDefault2["default"])(require("./../intl/da-DK.json"))["default"],
  "de-DE": (0, _interopRequireDefault2["default"])(require("./../intl/de-DE.json"))["default"],
  "en-US": (0, _interopRequireDefault2["default"])(require("./../intl/en-US.json"))["default"],
  "es-ES": (0, _interopRequireDefault2["default"])(require("./../intl/es-ES.json"))["default"],
  "fi-FI": (0, _interopRequireDefault2["default"])(require("./../intl/fi-FI.json"))["default"],
  "fr-FR": (0, _interopRequireDefault2["default"])(require("./../intl/fr-FR.json"))["default"],
  "hu-HU": (0, _interopRequireDefault2["default"])(require("./../intl/hu-HU.json"))["default"],
  "it-IT": (0, _interopRequireDefault2["default"])(require("./../intl/it-IT.json"))["default"],
  "ja-JP": (0, _interopRequireDefault2["default"])(require("./../intl/ja-JP.json"))["default"],
  "ko-KR": (0, _interopRequireDefault2["default"])(require("./../intl/ko-KR.json"))["default"],
  "nb-NO": (0, _interopRequireDefault2["default"])(require("./../intl/nb-NO.json"))["default"],
  "nl-NL": (0, _interopRequireDefault2["default"])(require("./../intl/nl-NL.json"))["default"],
  "pl-PL": (0, _interopRequireDefault2["default"])(require("./../intl/pl-PL.json"))["default"],
  "pt-BR": (0, _interopRequireDefault2["default"])(require("./../intl/pt-BR.json"))["default"],
  "ru-RU": (0, _interopRequireDefault2["default"])(require("./../intl/ru-RU.json"))["default"],
  "sv-SE": (0, _interopRequireDefault2["default"])(require("./../intl/sv-SE.json"))["default"],
  "tr-TR": (0, _interopRequireDefault2["default"])(require("./../intl/tr-TR.json"))["default"],
  "uk-UA": (0, _interopRequireDefault2["default"])(require("./../intl/uk-UA.json"))["default"],
  "zh-CN": (0, _interopRequireDefault2["default"])(require("./../intl/zh-CN.json"))["default"],
  "zh-TW": (0, _interopRequireDefault2["default"])(require("./../intl/zh-TW.json"))["default"]
};
var formatMessage = (0, _intl.messageFormatter)(intlMessages);
var STATUS_CANCEL = 'cancel';
exports.STATUS_CANCEL = STATUS_CANCEL;
var STATUS_COMPLETE = 'complete';
exports.STATUS_COMPLETE = STATUS_COMPLETE;

var Tour = (0, _autobindDecorator["default"])(_class = (_temp = _class2 =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2["default"])(Tour, _React$Component);

  function Tour(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Tour);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Tour).call(this, props));
    _this.state = {
      current: 0,
      total: 0,
      hidden: false
    };
    return _this;
  }

  (0, _createClass2["default"])(Tour, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this.props.children) {
        this.setState({
          total: _react["default"].Children.count(this.props.children)
        });
      }
    }
  }, {
    key: "onConfirm",
    value: function onConfirm() {
      if (this.state.current === this.state.total - 1) {
        this.setState({
          hidden: true
        });

        if (this.props.onTourEnd) {
          this.props.onTourEnd(STATUS_COMPLETE);
        }
      } else {
        this.setState({
          current: this.state.current + 1
        });
      }
    }
  }, {
    key: "onCancel",
    value: function onCancel() {
      this.setState({
        current: 0,
        hidden: true
      });

      if (this.props.onTourEnd) {
        this.props.onTourEnd(STATUS_CANCEL);
      }
    }
  }, {
    key: "onHide",
    value: function onHide(event, currentChild) {
      if (this.props.clickOutsideAction === 'next') {
        (0, _events.chain)(this.onConfirm, currentChild.props.onConfirm)(event);
      } else if (this.props.clickOutsideAction === 'skip') {
        (0, _events.chain)(this.onCancel, currentChild.props.onCancel)(event);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          children = _this$props.children,
          disableProgress = _this$props.disableProgress,
          clickOutsideAction = _this$props.clickOutsideAction;
      var _this$state = this.state,
          current = _this$state.current,
          total = _this$state.total,
          hidden = _this$state.hidden;

      if (!children || _react["default"].Children.count(children) <= this.state.current) {
        throw new Error('Tour should always have children!');
      }

      if (hidden) {
        return null;
      }

      var currentChild = _react["default"].Children.toArray(this.props.children)[current];

      return _react["default"].cloneElement(currentChild, {
        disableProgress: disableProgress,
        currentStep: current + 1,
        totalSteps: total,
        dismissible: clickOutsideAction === 'skip' || clickOutsideAction === 'next' ? true : undefined,
        onConfirm: (0, _events.chain)(this.onConfirm, currentChild.props.onConfirm),
        onCancel: (0, _events.chain)(this.onCancel, currentChild.props.onCancel),
        onHide: function onHide(event) {
          return _this2.onHide(event, currentChild);
        },
        confirmLabel: currentChild.props.confirmLabel || (current < total - 1 ? formatMessage('next') : formatMessage('done')),
        cancelLabel: currentChild.props.cancelLabel || (current < total - 1 ? formatMessage('skip') : null)
      });
    }
  }]);
  return Tour;
}(_react["default"].Component), (0, _defineProperty2["default"])(_class2, "propTypes", {
  /** Hide progress, the current step/total steps number in the
   *  top right of a coachmark, this is passed through */
  disableProgress: _propTypes["default"].bool,

  /** when not clicking on the coachmark itself, this determines the behavior.
   *  - 'next': a click off will behave the same as clicking 'next' on a coachmark
   *  - 'skip': a click off will skip the remainder of the tour
   *  - 'noop': a click off will do nothing
   **/
  clickOutsideAction: _propTypes["default"].oneOf(['next', 'skip', 'noop']),

  /**
   * Event when tour is finished. It's passed 'cancel' if the tour ended via the skip button
   * or other form of early dismissal. It's passed 'complete' if the tour is completed.
   */
  onTourEnd: _propTypes["default"].func
}), (0, _defineProperty2["default"])(_class2, "defaultProps", {
  disableProgress: false,
  clickOutsideAction: 'noop'
}), _temp)) || _class;

exports["default"] = Tour;
